File: app/src/main/java/com/reteno/sample/fragments/database/InteractionReadDialogFragment.java
package com.reteno.sample.fragments.database;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.transition.AutoTransition;
import androidx.transition.TransitionManager;

import com.reteno.core.data.local.database.manager.RetenoDatabaseManagerInteraction;
import com.reteno.core.data.local.model.interaction.InteractionDb;
import com.reteno.sample.R;
import com.reteno.sample.databinding.ItemDbInteractionBinding;

import java.util.List;
import java.util.Locale;


public class InteractionReadDialogFragment extends BaseReadDialogFragment<InteractionDb, ItemDbInteractionBinding, InteractionReadDialogFragment.InteractionViewHolder, InteractionReadDialogFragment.InteractionAdapter> {

    private RetenoDatabaseManagerInteraction databaseManager;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        databaseManager = serviceLocator.getRetenoDatabaseManagerInteractionProvider().get();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        databaseManager = null;
    }

    @Override
    protected void initAdapter() {
        adapter = new InteractionAdapter(() ->
                TransitionManager.beginDelayedTransition(bindingMain.rvList, new AutoTransition())
        );
    }

    @Override
    protected void initCount() {
        long count = databaseManager.getInteractionCount();
        bindingMain.tvCount.setText(String.format(Locale.US, "Count: %d", count));
    }

    @Override
    protected void initItems() {
        List<InteractionDb> newItems = databaseManager.getInteractions(null);
        adapter.setItems(newItems);
    }

    @Override
    protected void deleteItems(int count) {
        List<InteractionDb> interactions = databaseManager.getInteractions(count);
        for (InteractionDb interaction : interactions) {
            databaseManager.deleteInteraction(interaction);
        }
    }

    //==============================================================================================
    static class InteractionAdapter extends BaseReadAdapter<InteractionDb, ItemDbInteractionBinding, InteractionViewHolder> {

        InteractionAdapter(ViewHolderListener listener) {
            super(listener);
        }

        @NonNull
        @Override
        public InteractionViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            ItemDbInteractionBinding binding = ItemDbInteractionBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
            return new InteractionViewHolder(binding);
        }


        @Override
        protected void initListeners(ItemDbInteractionBinding binding) {
            binding.ivExpand.setOnClickListener(v -> {
                if (binding.llContent.getVisibility() == View.VISIBLE) {
                    onExpandCollapseClickListener.onExpandCollapse();
                    binding.llContent.setVisibility(View.GONE);
                    binding.ivExpand.setImageResource(R.drawable.ic_expand_more);
                } else {
                    onExpandCollapseClickListener.onExpandCollapse();
                    binding.llContent.setVisibility(View.VISIBLE);
                    binding.ivExpand.setImageResource(R.drawable.ic_expand_less);
                }
            });
        }
    }

    //==============================================================================================
    static class InteractionViewHolder extends BaseReadViewHolder<InteractionDb, ItemDbInteractionBinding> {

        InteractionViewHolder(ItemDbInteractionBinding binding) {
            super(binding);
        }

        @Override
        protected void bind(InteractionDb interaction) {
            bindingHolder.tvInteractionId.setText(interaction.getInteractionId());
            bindingHolder.tvStatus.setText(interaction.getStatus().toString());
            bindingHolder.tvTime.setText(interaction.getTime());
            bindingHolder.tvToken.setText(interaction.getToken());
        }
    }
}

File: app/src/main/java/com/reteno/sample/fragments/database/InteractionReadDialogFragment.kt
package com.reteno.sample.fragments.database

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.transition.AutoTransition
import androidx.transition.TransitionManager
import com.reteno.core.data.local.database.manager.RetenoDatabaseManagerInteraction
import com.reteno.core.data.local.model.interaction.InteractionDb
import com.reteno.sample.R
import com.reteno.sample.databinding.ItemDbInteractionBinding
import com.reteno.sample.fragments.database.InteractionReadDialogFragment.InteractionAdapter
import com.reteno.sample.fragments.database.InteractionReadDialogFragment.InteractionViewHolder
import java.util.Locale

internal class InteractionReadDialogFragment :
    BaseReadDialogFragment<InteractionDb, ItemDbInteractionBinding, InteractionViewHolder, InteractionAdapter>() {
    private var databaseManager: RetenoDatabaseManagerInteraction? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        databaseManager = serviceLocator.retenoDatabaseManagerInteractionProvider.get()
    }

    override fun onDestroy() {
        super.onDestroy()
        databaseManager = null
    }

    override fun initAdapter() {
        adapter = InteractionAdapter {
            TransitionManager.beginDelayedTransition(
                bindingMain!!.rvList, AutoTransition()
            )
        }
    }

    override fun initCount() {
        val count = databaseManager!!.getInteractionCount()
        bindingMain!!.tvCount.text = String.format(Locale.US, "Count: %d", count)
    }

    override fun initItems() {
        val newItems = databaseManager!!.getInteractions(null)
        adapter!!.setItems(newItems)
    }

    override fun deleteItems(count: Int) {
        val interactions = databaseManager!!.getInteractions(count)
        for (interaction in interactions) {
            databaseManager!!.deleteInteraction(interaction)
        }
    }

    //==============================================================================================
    internal class InteractionAdapter(listener: ViewHolderListener) :
        BaseReadAdapter<InteractionDb, ItemDbInteractionBinding, InteractionViewHolder>(listener) {
        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): InteractionViewHolder {
            val binding =
                ItemDbInteractionBinding.inflate(LayoutInflater.from(parent.context), parent, false)
            return InteractionViewHolder(binding)
        }

        override fun initListeners(binding: ItemDbInteractionBinding) {
            binding.ivExpand.setOnClickListener {
                if (binding.llContent.visibility == View.VISIBLE) {
                    onExpandCollapseClickListener.onExpandCollapse()
                    binding.llContent.visibility = View.GONE
                    binding.ivExpand.setImageResource(R.drawable.ic_expand_more)
                } else {
                    onExpandCollapseClickListener.onExpandCollapse()
                    binding.llContent.visibility = View.VISIBLE
                    binding.ivExpand.setImageResource(R.drawable.ic_expand_less)
                }
            }
        }
    }

    //==============================================================================================
    class InteractionViewHolder(binding: ItemDbInteractionBinding) :
        BaseReadViewHolder<InteractionDb, ItemDbInteractionBinding>(binding) {
        override fun bind(model: InteractionDb) {
            bindingHolder.tvInteractionId.setText(model.interactionId)
            bindingHolder.tvStatus.setText(model.status.toString())
            bindingHolder.tvTime.setText(model.time)
            bindingHolder.tvToken.setText(model.token)
        }
    }
}
